Finish refactor of UID/GID usage to a new struct

corhere opened the pull request:
**- What I did**
Finished the refactoring which was started in #33362: replace all remaining usages of pairs of `[]idtools.IDMap` with `idtools.IdentityMapping`.

**- How I did it**
To ensure that there would be no change in semantics with uninitialized fields, I replaced `[]IDMap` struct fields with `IdentityMapping` values and refactored all code to pass `IdentityMapping` structs by value. BuildKit makes heavy use of pass-by-reference `*IdentityMapping` values; I handled that as a special case in the builder-next adapters.

Changing the `IdentityMapping` associated functions from pointer receivers to value receivers is a non-breaking change as the pointer-receiver version of the functions would also panic when called on a nil pointer. The only notable difference would be in the stack trace: the nil dereference will now occur in the implicit dereference at the call site, rather than inside the implementation.

**- How to verify it**
CI

**- Description for the changelog**
N/A

**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
Starting do do a first review; w.r.t. the first commit; perhaps it's still possible to change (if it's only buildkit consuming it); I'll let @tonistiigi and @crazy-max comment on that;

> IdentityMapping values are already used extensively within BuildKit so
we cannot do that without also refactoring BuildKit in lockstep.

thaJeztah commented on daemon/graphdriver/aufs/aufs.go:
```diff
@@ -461,8 +460,7 @@ func (a *Driver) DiffGetter(id string) (graphdriver.FileGetCloser, error) {
 	return chrootarchive.UntarUncompressed(diff, path.Join(a.rootPath(), "diff", id), &archive.TarOptions{
-		UIDMaps: a.uidMaps,
-		GIDMaps: a.gidMaps,
+		IDMap: idtools.NewIDMappingsFromMaps(a.uidMaps, a.gidMaps),
```
Wondering why you're introducing the deprecated utility in these intermediate commits, before replacing them later, or was that just to make it easier to find back which locations had to be updated? https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/pkg/idtools/idtools.go#L119

thaJeztah commented on pkg/idtools/idtools.go:
```diff
@@ -165,35 +159,37 @@ func (i *IdentityMapping) ToContainer(pair Identity) (int, int, error) {
-// TODO: remove this once everything has been refactored to use pairs
+// UIDs returns the mapping for UID.
+//
+// Deprecated: use the UIDMap field directly.
```
Slightly wondering if this would still be useful / convenient to have (to handle the nil cases), but I'd have to check again where / how it's all used (otherwise, good with me to clean it up 👍)

corhere replied:
Oh yeah, the nil cases... 🤦 I made a mistake. There's at least one case where an `*IdentityMapping` function argument is immediately dereferenced without any nil-pointer checks. That's no good; excuse me while I go revise this PR.

thaJeztah replied:
Oh! Good that I commented then (probably shouldn't have "LGTM'd while on the road 🤦); no worries!

thaJeztah reviewed (approved):
nice! overall this looks quotes straightforward (and looks good to me); left one comment about the intermediate commits, but not too tied to that (just wondering if we could avoid the extra code churn)

corhere committed:
Finish refactor of UID/GID usage to a new struct

Finish the refactor which was partially completed with commit
34536c498d56, passing around IdentityMapping structs instead of pairs of
[]IDMap slices.

Existing code which uses []IDMap relies on zero-valued fields to be
valid, empty mappings. So in order to successfully finish the
refactoring without introducing bugs, their replacement therefore also
needs to have a useful zero value which represents an empty mapping.
Change IdentityMapping to be a pass-by-value type so that there are no
nil pointers to worry about.

The functionality provided by the deprecated NewIDMappingsFromMaps
function is required by unit tests to to construct arbitrary
IdentityMapping values. And the daemon will always need to access the
mappings to pass them to the Linux kernel. Accommodate these use cases
by exporting the struct fields instead. BuildKit currently depends on
the UIDs and GIDs methods so we cannot get rid of them yet.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented on daemon/graphdriver/fuse-overlayfs/fuseoverlayfs.go:
```diff
@@ -175,22 +172,12 @@ func (d *Driver) Create(id, parent string, opts *graphdriver.CreateOpts) (retErr
-	if err != nil {
-		return err
-	}
-	root := idtools.Identity{UID: rootUID, GID: rootGID}
-
-	dirID := idtools.Identity{
-		UID: rootUID,
-		GID: rootGID,
-	}
```
I noticed that `root == dirID` in this driver so I coalesced the two when I refactored it. But I noticed that other drivers follow a very similar pattern, except that their `root != dirID`: they set `dirID.UID` to `idtools.CurrentIdentity().UID`. Could this be a bug in the fuseoverlayfs driver?

thaJeztah replied:
^^ @AkihiroSuda

AkihiroSuda replied:
Yes, seems a bug

thaJeztah replied:
Thanks for looking! Ok, given that this PR should not make the situation "worse"; let's fix that separately.

@corhere do you want to work on that? (Otherwise feel free to create a tracking issue for it)

thaJeztah commented:
@AkihiroSuda @tonistiigi PTAL 🙏

thaJeztah reviewed (approved):
still LGTM, thanks!
